### PR TITLE
feat(node): integrated pagination for devices list view

### LINKED ISSUE #590: Node / Devices pagination
> Note: update Protobuf

### PR DESCRIPTION
**Pull request by vladimirnikolic509**
Fixes #590

### CODE REVIEW

File: pkg/graph/devices.go [hunk 1]
```diff
@@ -436,6 +436,14 @@ func (c *DevicesController) List(ctx context.Context, req *connect.Request[pb.Qu
 		ctx = WithNamespaceFilter(ctx, q.GetNamespace())
 	}
 
+	limit := q.GetLimit()
+	page := q.GetPage()
+
+	if limit != 0 && page != 0 {
+		ctx = WithLimit(ctx, int(limit))
+		ctx = WithSkip(ctx, (int(page)-1)*int(limit))
```
Comment (slntopp) [inline pkg/graph/devices.go:444]: Should be named WithOffset

File: pkg/graph/common.go [hunk 2]
```diff
@@ -243,6 +243,27 @@ func (r *infinimeshCommonActionsRepo) AccessLevelAndGet(ctx context.Context, log
 	return nil
 }
 
+const listObjectsOfKindWithPagination = `
```
Comment (slntopp) [inline pkg/graph/common.go:246]: it'd be better to embed `LIMIT` clause into the query same way as it's done with Namespace filter

**Review by slntopp — CHANGES_REQUESTED** [review]
I suggest to add Count method/field as well, so we don't have unlimited amout of pages 

Additionally, as a part of ongoing refactoring and refinement, please provide unit tests for `ListQuery` and update `DevicesService.List` tests to cover the pagination functionality

File: pkg/graph/devices.go (hunk 1 above)
Comment (vladimirnikolic509) [reply pkg/graph/devices.go]: Fixed

File: pkg/graph/common.go (hunk 2 above)
Comment (vladimirnikolic509) [reply pkg/graph/common.go]: Fixed

**Review by slntopp — CHANGES_REQUESTED** [review]
I think count doesn't worth a separate query, Arango should be using lazy queries, so you can probably just get them all and then slice the page out of the result without any or big performance overhead

File: pkg/graph/devices.go [hunk 3]
```diff
@@ -479,9 +467,23 @@ func (c *DevicesController) List(ctx context.Context, req *connect.Request[pb.Qu
 		r = append(r, &dev)
 	}
 
+	filteredDevices := r
+
+	if limit > 0 && page > 0 {
```
Comment (slntopp) [inline pkg/graph/devices.go:472]: this condition should be part of the query, i.e. the query goes like

1. Get all accessible devices
2. `LET count = COUNT(result)`
3. `%s` add pagination here if requested
4. RETURN { result, count }

File: pkg/graph/devices.go [hunk 4]
```diff
@@ -479,9 +467,23 @@ func (c *DevicesController) List(ctx context.Context, req *connect.Request[pb.Qu
...
+		} else {
+			filteredDevices = r[start:end]
+		}
+	}
+
 	return connect.NewResponse(&devpb.Devices{
-		Devices: r,
-		// Count:   count,
+		Devices: filteredDevices,
+		// Total:   len(r),
```
Comment (slntopp) [inline pkg/graph/devices.go:486]: and with accordance to the previous comment, total should be returned

File: pkg/graph/devices_test.go [hunk 5]
```diff
@@ -619,11 +620,64 @@ func TestMakeDevicesToken_Success(t *testing.T) {
 	assert.NotNil(t, res)
 }
 
-func TestList_Success(t *testing.T) {
+func TestList_Error_1(t *testing.T) {
```
Comment (slntopp) [inline pkg/graph/devices_test.go:623]: Name tests like `TestFunction_FailsOn_WhatItisFailingOn`
For example this test could be named `TestList_FailsOn_ListQuery`

File: pkg/graph/devices.go (hunk 3 above)
Comment (slntopp) [reply pkg/graph/devices.go:472]: So it should be part of `ListQuery` logic

**Review by slntopp — CHANGES_REQUESTED** [review]

File: pkg/graph/devices.go (hunk 3 above)
Comment (vladimirnikolic509) [reply pkg/graph/devices.go:472]: Then we will have to move the logic of reading from cursor too on ListQuery. @slntopp

Comment (slntopp) [reply pkg/graph/devices.go:472]: Why is that?

Comment (slntopp) [reply pkg/graph/devices.go:472]: Pagination should be implemented inside the AQL query, same as filter, so cursor can stay where-ever it is right now. 

Ideally ofcourse we could introduce generic, but Protobuf structures don't provide `Set` methods unfortunately

Comment (vladimirnikolic509) [reply pkg/graph/devices.go:472]: https://github.com/infinimesh/infinimesh/pull/671#pullrequestreview-1860136750
> I think count doesn't worth a separate query, Arango should be using lazy queries, so you can probably just get them all and then slice the page out of the result without any or big performance overhead

You said to use single query and then get total count from it, and slice the result as per sent value of limit/page. At least that’s what I understood.
Then if I am to return count from ListQuery itself, how do I return count from ListQuery without reading from cursor in ListQuery function itself?
@slntopp

Comment (slntopp) [reply pkg/graph/devices.go:472]: You can do it all withing AQL as i've described in the first comment of this thread.
Your base query would be smth like this:

```aql
LET result = (
FOR node, edge, path IN 0..@depth OUTBOUND @from
GRAPH @permissions_graph
OPTIONS {order: "bfs", uniqueVertices: "global"}
FILTER IS_SAME_COLLECTION(@@kind, node)
FILTER edge.level > 0
%s
    LET perm = path.edges[0]
    LET last = path.edges[-1]
	RETURN MERGE(node, {
	    uuid: node._key,
	    access: {
	        level: last.role == 2 ? last.level : perm.level,
	        role:  last.role == 2 ? last.role : perm.role,
	        namespace: last.role == 2 ? null : path.vertices[-2]._key
	     }
	    }
    )
)

LET count = LENGTH(result)

%s
```

And the second `%s` would be replaced with `RETURN {result, count}` by default. Unless, there are pagination parameters provided, then you would replace it with smth like:

```aql
LET paginated = SLICE(result, offset, offset + page_length)
RETURN {result: paginated, count}
```

Please manually test the queries with the local or dev ArangoDB deployment

Comment (slntopp) [reply pkg/graph/devices.go:472]: > note on this, would be great to replace simple fmt.Sprintf with some sort of templating or at least named replacements.

Comment (slntopp) [reply pkg/graph/devices.go:472]: > Another note, you can use lazy evaluation to define this condition, like 
```
RETURN { 
 result: offset != NULL && limit != NULL ? SLICE(...) : result,
 count
}
```

Comment (vladimirnikolic509) [reply pkg/graph/devices.go]: Fixed

File: pkg/graph/devices_test.go (hunk 5 above)
Comment (vladimirnikolic509) [reply pkg/graph/devices_test.go]: Fixed

File: pkg/graph/devices.go (hunk 4 above)
Comment (vladimirnikolic509) [reply pkg/graph/devices.go]: Fixed

**Comment by slntopp** [comment]
- Created a separate GenericRepo for ListQuery and other CRUD operations in the future.
- Updated tests
- Fixed AQL Query to make it possible to limit result with 0 offset

**Comment by slntopp** [comment]
- renamed field `page` -> `offset`

**Review by slntopp — APPROVED** [review]

**Pull request merged**